csharp/src/Ice/ConnectionI.cs

```diff
@@ -1647,31 +1648,40 @@ private void setState(int state)
                 case StateActive:
                 {
                     //
-                    // Can only switch from holding or not validated to
-                    // active.
+                    // Can only switch to active from holding or not validated.
                     //
                     if (_state != StateHolding && _state != StateNotValidated)
                     {
                         return;
                     }
+
+                    if (_maxDispatches > 0 && _dispatchCount >= _maxDispatches)
+                    {
+                        // Don't resume reads if the max dispatch count is reached.
+                        return;
+                    }
+
                     _threadPool.register(this, SocketOperation.Read);
                     break;
                 }
 
                 case StateHolding:
                 {
                     //
-                    // Can only switch from active or not validated to
-                    // holding.
+                    // Can only switch to holding from active or not validated.
                     //
                     if (_state != StateActive && _state != StateNotValidated)
                     {
                         return;
                     }
-                    if (_state == StateActive)
+
+                    if (_maxDispatches > 0 && _dispatchCount >= _maxDispatches)
                     {
-                        _threadPool.unregister(this, SocketOperation.Read);
+                        // Reads are already disabled if the max dispatch count is reached.
+                        return;
                     }
+
+                    _threadPool.unregister(this, SocketOperation.Read);
```

Comment:
There is an extra change compared to the old code:
 - old code, state = StateNotValidated, we don't unregister and just 'break'
 - new code, state = StateNotValidated treated like StateActive

 Is this intentional?

Reply:
It doesn't make a difference. If `state == StateNotValidated`, unregistering reads is fine, it will be a no-op. The previous code assumes reads were already unregistered -- it wasn't bogus because it's indeed the case.

I find it clearer to have the same code as for the `StateActive` case.

Reply:
Ok